docs/flags.md

```diff
@@ -0,0 +1,130 @@
+
+groups = {
+  'consumer': ['a', 'b', 'c'],
+  'enterprise': ['a', 'b', 'd'],
+}
+
+
+
+
+
+
+
+
+
+
+
+
+flags.exclude_all_resources()
+for arg in cfg['args']:
+  if arg in groups:
+    for r in groups[arg]:
+      flags.include_resource(r)
+  else:
+    flags.include_resource(r)
+```
+
+By making the flag sticky (the default), Tilt will record the flag between runs.
+
+
+### Specify services to edit
+You have many services that your users frequently need to run, but don't expect to edit. Your YAML includes images from your CI that are suitably recently (and match what's in production or staging).
+
+#### Command-Lines
+* `tilt up`: run all services, editing none
+* `tilt up -- --to-edit b`: run all services, editing b
+* `tilt up -- consumer --to-edit b --to-edit c`: run consumer services and setup b and c for editing.
+
+#### Tiltfile
+```python
+flags.define_string_list("to-run", args=True)
+flags.define_string_list("to-edit")
+cfg = flags.parse()
+
+# omitting the code from above Tiltfile that includes correct resources
+
+# only configure the build for services we want to edit
+if 'a' in cfg['to-edit']:
+  docker_build('a', './a')
+if 'b' in cfg['to-edit']:
+  docker_build('b', './b')
+if 'c' in cfg['to-edit']:
+  docker_build('c', './c')
+if 'd' in cfg['to-edit']:
+  docker_build('d', './d')
+```
+
+## How Flags Work
+Tilt stores flags in a file next to the Tiltfile called `tilt_flags.json`. When the Tiltfile executes `flags.parse()`, it loads that file as JSON. If it's the first execution after a Tilt up, the execution will merge the flags passed on the command-line with the values from the file (using the definition of flags as a guide) and write the new values to the file if necessary.
+
+Any change to the flags file will cause Tilt to reexecute the Tiltfile. You can modify the flags file with an editor or write scripts that write values into it, and Tilt will update.
+
+A call to `tilt flags` will connect to the running Tilt instance on the server port and cause Tilt to reexecute the Tiltfile, using the passed command-line args as if it's the first execution after a Tilt up.
+
+## Future Directions
+This section describes places we expect flags to go. Let us know if any of these would be particularly helpful to you and your team.
+
+### More Kinds of Flags
+You should be able to define more kinds of flags. For example, a list of ints or floats, or a singly-valued string. By moving error-checking to the built-in library, you can make your Tiltfile shorter and more correct.
+
+### Better Flag Manipulation
+Your users should be able to understand what reasonable values are. This could be from help output, tab-completion, or in the Web UI.
+
+Your users should be able to modify flags in the Web UI, without having to know there's a file underneath.
+
+### Escape Hatch for Flag Parsing
+You should be able to use a custom syntax for your flags. The flags module could take a function that would parse a command-line into a JSON value.
```

Comment:
FWIW, I think we don't even need this much - we just need to provide a `flags()` function that returns a raw array of strings, and then the user can do whatever they want to convert that to `cfg`.

Or maybe merging complicates that, I'm still confused by how merging works.

Reply:
Yeah, let's discuss a bit more?